AdminUI

Daenks opened the pull request:
![AdminUI Demo](http://i.imgur.com/XUINl71.jpg)

DebugUI has a new tab "AdminUI"

  Currently works just like the  $ window, type a command, see the output.
  Launches New DebugUI.ObjectEditor Form from parsed return data of "show object X" commands.
  DebugUI.ObjectEditor Form for displaying and eventually editing AdminObjects. 

New namespace Meridian59.Data.Models.AdminData:
  class AdminObject represents the object after being parsed, contains list of AdminObjectProperty
  class AdminObjectProperty represents a property of an AdminObject, has a name, tag, value.

DataController handles the List of AdminObjects. this still feels sloppy. I am considering writing AdminObjectList and AdminObjectPropertyList to bind everything together easier.

Daenks committed:
Not working, controller class just started. Copying HandleAdminMessage code back to DebugUI funtion of the same name works fine.

Daenks committed:
Applied ShaKrune's recommendations for appropriately using his data model.
Working code state, can send admin commands via text console.
"Show Object" command results sucessfully parsed and displayed.

cyberjunk commented on Meridian59.DebugUI/ObjectEditor.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public partial class ObjectEditor : Form
+    {
+        public ObjectEditor(AdminObject obj)
```
Please do not pass the datasource as a constructor parameter.

Instead define a property "DataSource" on your UI (view), according to MVC pattern the type of this property of course would be the one of the according underlying model providing the data for the view (your "AdminData" class for example if you need all of it - or in this case probably just "AdminObject")

Please have a look at "RoomInfoView.cs" or "RoomObjectsView.cs" for examples for "DataSource" property.
You can adopt their definition of a "DataSource" property. Don't forget the [DesignerSerialization...] attributes above the signature.

Daenks replied:
This is not completed yet. There are a lof of ways I can work better within your design/architecture. I'll work on that slowly. I feel like now that I have AdminData class in Meridian59 I might need an AdminUIController in DebugUI that handles talking to AdminData.

I probably need to define some Interfaces for AdminObject and AdminProperty, INotifyProeprtyChanged is probably not going to work in the long run.

cyberjunk commented on Meridian59/Data/DataController.cs:
```diff
@@ -1902,6 +1913,52 @@ protected void HandleMessage(MessageMessage Message)
             ChatMessages.Add(Message.Message);           
         }
 
+        public void HandleAdminMessage(AdminMessage Message)
```
This should be 'protected' like the other internal handlers, because it is only called from class internal _HandleIncomingGameModeMessage(GameMessage Message)_

Daenks replied:
fixed

cyberjunk commented on Meridian59/Meridian59.csproj:
```diff
@@ -561,7 +565,9 @@
   </ItemGroup>
-  <ItemGroup />
+  <ItemGroup>
+    <Folder Include="Protocol\GameMessages\AdminMode\" />
```
I don't think this folder "AdminMode" should be there.

AFAIK the network protocol only knows "LoginMode" - the messages with "AP_" prefix from proto.h and "GameMode", the messages with "BP_" prefix from proto.h - or am I missing something?

cyberjunk commented on Meridian59.DebugUI/CustomDataGridColumns/TagColumn.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Meridian59.DebugUI.CustomDataGridColumns
+{
+    class TagCell : DataGridViewTextBoxCell
```
Does this class have any purpose yet or is it a preparation for something you have in mind ?

Daenks replied:
Just a placeholder. I want to have a custom column for the "Tag/Blakserv DataType" column in the ObjectEditor form.

cyberjunk commented on Meridian59.DebugUI/DebugForm.cs:
```diff
@@ -85,6 +91,17 @@ public DebugForm()
+            oe.Show();
+        }
+
+        public void HandleAdminMessage(AdminMessage Message)
```
no need for this empty method anymore?

Daenks replied:
This event still exists in the latest version with a new name, it ties the debug form to the AdminData object for showing a new form when needed.

cyberjunk commented on Meridian59/Data/DataController.cs:
```diff
@@ -1902,6 +1913,52 @@ protected void HandleMessage(MessageMessage Message)
+                if (obj != null)
+                {
+                    obj.SetProperties(props);
+                    AdminData.WatchObject(obj);
```
I'm not exactly sure if I get this WatchObject() call and the event correctly, but I think I would prefer if:

(a) Instead your AdminData class would have a property named like "WatchedObject" of type "AdminObject".
This property would represent the currently "watched" object/the data that should be shown in something like "detailview". Your model/datacontroller code then sets this property, like "AdminData.WatchObject = obj;". Your UI code should listen for the "PropertyChanged" event of "AdminData" class.

(b) So ff not already so, your AdminData class should implement the "INotifyPropertyChanged" event (see all the other models for how to implement properties then raising PropertyChanged event).

Daenks replied:
AdminObject and AdminObjectProperty are currently implementing the INotifyPropertyChanged interface. AdminData is tracking a set of objects at a time. There is an event sent to open the window after the object is "Tracked" (new terminology, replaced watched). This will also be used in the future to poll the server for updates of tracked objects (to refresh the UI). 

There's maybe a better way and I'm eager to hear it.

Daenks committed:
Applied ShaKrune's recommendations for appropriately using his data model.
Working code state, can send admin commands via text console.
"Show Object" command results sucessfully parsed and displayed.

Daenks committed:
added adminobjectproperty constructor /w init values

Daenks committed:
Merge branch 'AdminUI' of https://github.com/Daenks/meridian59-dotnet into AdminUI

Daenks committed:
fixed OnWatch message in DebugForm

Daenks committed:
Merge branch 'master' of https://github.com/Daenks/meridian59-dotnet into AdminUI

Conflicts:
	Meridian59.DebugUI/Meridian59.DebugUI.csproj
	Meridian59/Data/DataController.cs
	Meridian59/Meridian59.csproj

Daenks committed:
fix project issue

Daenks committed:
Working multi-object viewing and editing.

DataController:
  Moved Handle....() from DataController to AdminData
AdminData:
  Renamed "watched" object list, functions, etc to "tracked"
  added PacketSend event
  HandleAdminMessage() moved here
  HandleAdminShowObjectMessage():
    resolved weird regex logic
    fixed regex to properly parse RESOURCEs
    now handles objects that already exist on the tracked list
AdminObject:
  Properties's PropertyChanged events are subscribed to
  Sends its own PropertyChanged event when needed
AdminObjectProperty:
  PropertyValue changed to private-backed field to add set logic to raise PropertyChanged events.
  added constructor that sets PropertyChanged event subscribers
  constructors do not trigger PropertyChanged by setting the backing field
  Added GetOwner()
  Clean up
ObjectEditor:
  dataGridView now uses a BindingSource to properly send changes to the list
  RefreshData() to redraw the dataGridView on change
  event handler to call RefreshData()
  GetTrackedObject()
DebugForm:
  added event handler for passing packets on to datacontroller
  added event handler for removing tracked objects on window close

Daenks committed:
Working multi-object viewing and editing.

DataController:
  Moved Handle....() from DataController to AdminData
AdminData:
  Renamed "watched" object list, functions, etc to "tracked"
  added PacketSend event
  added LogAdminMessage event to pass text back to AdminUI Log
  HandleAdminMessage() moved here
  HandleAdminShowObjectMessage():
    resolved weird regex logic
    fixed regex to properly parse RESOURCEs
    now handles objects that already exist on the tracked list
AdminObject:
  Properties's PropertyChanged events are subscribed to
  Sends its own PropertyChanged event when needed
AdminObjectProperty:
  PropertyValue changed to private-backed field to add set logic to raise PropertyChanged events.
  added constructor that sets PropertyChanged event subscribers
  constructors do not trigger PropertyChanged by setting the backing field
  Added GetOwner()
  Clean up
ObjectEditor:
  dataGridView now uses a BindingSource to properly send changes to the list
  RefreshData() to redraw the dataGridView on change
  event handler to call RefreshData()
  GetTrackedObject()
DebugForm:
  added event handler for passing packets on to datacontroller
  added event handler for removing tracked objects on window close

Daenks committed:
Fixing last commit

Daenks committed:
Merge branch 'AdminUI' of https://github.com/Daenks/meridian59-dotnet into AdminUI

Conflicts:
	Meridian59.DebugUI/ObjectEditor.cs
	Meridian59/Data/Models/AdminData/AdminData.cs

Daenks committed:
changed DataController.HandleAdminMessage() to private

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminObject.cs:
```diff
@@ -0,0 +1,72 @@
+            }
+            else
+            {
+                Properties = new BaseList<AdminObjectProperty>();
```
I prefer initialising collection instances once and define them readonly, to make sure they are correctly cleared/etc. rather than overwritten with a new instance (important for databinding!).

To achieve this, you should add the following field declaration to your class:

> > protected readonly BaseList< AdminObjectProperty > properties = new BaseList<..>();

Then return this instance in the getter of your:

> > public BaseList< AdminObjectProperty > Properties { get { return properties; } } 

In your constructor, you would then have something like:

> > public AdminObject(string classname, int objectnumber, IEnumerable< AdminObjectProperty > props = null)

And in its implementation, you add the items to your own collection:

> > for(.... in props) ..[addtoownpropertieslist]

cyberjunk commented on Meridian59.DebugUI/ObjectEditor.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Meridian59.DebugUI
+{
+    public partial class ObjectEditor : Form
```
Usually the classname of all views refers to the model name that is shown in the view, e.g.:
- Models.RoomInfo --> RoomInfoView
- Models.ObjectBase --> ObjectBaseView

It looks like a proper name would be "AdminObjectView", because this class seems to display the data of the "AdminObject" model.

Daenks replied:
Makes sense.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+    public class AdminData : INotifyPropertyChanged, IClearable
+    {
+        //Handles the objects we're watching in the AdminUI
+        private BaseList<AdminObject> TrackedObjects { get; set; }
```
Have a look at the comments somewhere else about creating instances of internal collections, which are ment for databinding, exactly once at class initialisation by using the 'readonly' modifier (I know I haven't updated all my own code to this yet, but it's the way to go).

> > protected readonly BaseList< AdminObject > trackedObjects = new BaseList< AdminObject >();
> > public BaseList< AdminObject > TrackedObjects { get { return trackedObjects; } }

Daenks replied:
Will do.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <summary>
+        /// Interface to the Client's Network
+        /// </summary>
+        public event GameMessageEventHandler PacketSend;
```
I'm not sure about this, construct/sending game-messages from model layer.
At least this would be the first model class in Data.Models which actually tries to send a gamemessage.

Daenks replied:
Yea, i think this logic needs to go back to the DataController, along with the HandleAdminSTUFF() functions.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+        }
+        #endregion
+
+        void adminObjectProperty_ValueChanged(object sender, PropertyChangedEventArgs e)
```
The current style for event handlers in the core library looks like this (I know especially the debug lib project is ugly formatted and does not stick to it):

> > OnInstanceEventName(...)

So this should be like (please always have visibility keyword, by default protected or public):

> > protected void OnAdminObjectPropertyChanged(...)

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+        }
+        #endregion
+
+        public void HandleAdminMessage(AdminMessage message)
```
It might be a bit uncommon, but I try to write the parameters with a capital start letter.

> > public void HandleAdminMessage(AdminMessage Message)

Daenks replied:
ReSharper keeps yelling at me about your standard haha. I need to set the options for this solution :)

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+            var matches = regex.Matches(message.Message);
+            if (matches.Count != 1)
+            {
+                throw new Exception("Improper Response in HandleAdminShowObjectMessage()");
```
Please avoid throwing exceptions, this would crash the client.
You probably should create a notification or log etc.

Daenks replied:
Absolutely. I won't throw any in a finished product. Useful for debugging and troubleshooting.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminObject.cs:
```diff
@@ -0,0 +1,72 @@
+    /// <summary>
+    /// Object that represents a displayed object in the Administration UI
+    /// </summary>
+    public class AdminObject : INotifyPropertyChanged
```
I don't think this class correctly implements INotifyPropertyChanged.
Someone could change the vlaues for "ObjectNumber" or "ClassName" and they would not raise changed events.

To implement INotifyPropertychanged:
(1) Add constants for your property names to the beginning of the class
(2) Create protected fields, not only properties. Provide the fieldvalue in the getter of the property.
(3) Implement the setter of the property to check the value and if different, set the field and raise the propertychanged event.

Example for (3):

> > public uint NameRID
> > {
> >             get { return nameRID; }
> >             set
> >             {
> >                 if (nameRID != value)
> >                 {
> >                     nameRID = value;
> >                     RaisePropertyChanged(new PropertyChangedEventArgs(PROPNAME_NAMERID));
> >                 }
> >             }
> >         }

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminObject.cs:
```diff
@@ -0,0 +1,72 @@
+        }
+
+        //so later we can do thing upon property update
+        public void SetProperties(BaseList<AdminObjectProperty> properties)
```
This code should be in the set { } implementation of the property.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminObjectProperty.cs:
```diff
@@ -0,0 +1,67 @@
+        public string PropertyName { get; set; }
+        public string PropertyType { get; set; }
+        
+        private string _propertyValue { get; set; }
```
Please avoid using "_" prefixes. Start the property name with a capital letter, and the field name with a non-capital one. Also by default use "protected" rather than "private" - unless you have reasons to, because it fits more cases by default IMO.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+        public void TrackAdminObject(AdminObject adminObject)
+        {
+            TrackedObjects.Add(adminObject);
+            OnTrackedAdminObjectAdded(new AddTrackedAdminObjectEventHandlerArgs(this, adminObject));
```
You should not need an event like "AdminObjectAdded".
The BaseList class of "TrackedObjects" already provides an event named "ListChanged" for observing it.

Just subscribe this with an handler and check for type "ItemAdded".
This handler will be triggered by the line:

> > TrackedObjects.Add(adminObject);

Have a look at this example:
1) DataController subscribes "ListChanged" in its construtctor:

> > ChatMessages.ListChanged += OnChatMessagesListChanged;

2) Then wen someone does "ChatMessages.Add()" (or delete,..), it will run this code in "OnChatMessagesListChanged"

> > switch (e.ListChangedType)
> >             {
> >                 case ListChangedType.ItemAdded:
> >                     Logger.LogChat(ChatMessages[e.NewIndex].FullString);
> >                     break;
> >             }

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+
+        void adminObjectProperty_ValueChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (sender.GetType() == typeof (AdminObjectProperty))
```
I'm usually using the "is" keyword to check class, however there is a difference to your variant here.
This expression will probably return "false" for a class deriving from 'AdminObjectProperty'?
If this is what you want, it's OK. But usually you would "accept" any "more specialized" class too.
So this could be:

> > if (sender is AdminObjectProperty)

Daenks replied:
I didn't even know that was a keyword in C#.
TIL.
Will change, thanks. It wouldn't matter I think in this case, (as nothing should ever inherit AdminObjectProperty) but it won't hurt do it it right!

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminObjectProperty.cs:
```diff
@@ -0,0 +1,67 @@
+            get { return _propertyValue; }
+            set
+            {
+                _propertyValue = value;
```
You better should check equality of values before setting them when using the INotifyPropertyChanged interface.

Sometimes this can even be important to break out of a 2-way-binding loop:
Assume a user enters a new value into the view, the view updates the model-value by databinding, the model raises a changedevent (2-way-binding), the view updates its value again, which wlll make the model update, which raises changeevent and so on. To break this loop either the view or the model must check for an actually different value. The 'core' can't assume that what's building on top of it is corrrect, so please check in all models implementing INotifyPropertyChanged.

Daenks replied:
Yea my workaround for this binding loop resulted in some other code you have already commented on. I'll fix it. Thanks.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminObjectProperty.cs:
```diff
@@ -0,0 +1,67 @@
+            set
+            {
+                _propertyValue = value;
+                RaisePropertyChanged(new PropertyChangedEventArgs(_owner.ObjectNumber.ToString()));
```
The argument of the 'PropertyChangedEventArgs' must be the name of the Property which changed its value.
In this case it would be:

> > RaisePropertyChanged(new PropertyChangedEventArgs("PropertyValue"));

I usually define all the property names as constants. This is also good to check against the property name in the event handlers then.

cyberjunk commented on Meridian59/Data/Models/AdminData/AdminData.cs:
```diff
@@ -0,0 +1,153 @@
+            TrackedObjects.Add(adminObject);
+            OnTrackedAdminObjectAdded(new AddTrackedAdminObjectEventHandlerArgs(this, adminObject));
+        }
+        public void UntrackAdminObject(AdminObject adminObject)
```
A bit like with the comments above...

Instead of providing a "TrackAdminObject()" and "UntrackAdminObject()" method, observe the trackedObjects collection using its "ListChanged" event which you subscribe to in the constructor here.

Then from outside simply do:

> > Data.AdminData.TrackedObjects.Add(...)
> > Data.AdminData.TrackedObjects.Remove(..)

This way you're subscribed handlers will be called.
It's like hooking in, rather than forcing someone to call 'your' methods.
